cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -45,67 +49,109 @@ constexpr static inline const float kAdjustCentersWeight = 7.0f;
  * NB: no minibatch splitting is done here, it may require large amount of temporary memory (n_rows
  * * n_cluster * sizeof(float)).
  *
+ * @tparam LabelT label type
+ *
  * @param handle
  * @param[in] centers a pointer to the row-major matrix of cluster centers [n_clusters, dim]
  * @param n_clusters number of clusters/centers
  * @param dim dimensionality of the data
  * @param[in] dataset a pointer to the data [n_rows, dim]
+ * @param[in] dataset_norm pointer to the precomputed norm (for L2 metrics only) [n_rows]
  * @param n_rows number samples in the `dataset`
  * @param[out] labels output predictions [n_rows]
  * @param metric
  * @param stream
  * @param mr (optional) memory resource to use for temporary allocations
  */
+template <typename IdxT, typename LabelT>
 inline void predict_float_core(const handle_t& handle,
                                const float* centers,
                                uint32_t n_clusters,
                                uint32_t dim,
                                const float* dataset,
+                               const float* dataset_norm,
                                size_t n_rows,
-                               uint32_t* labels,
+                               LabelT* labels,
                                raft::distance::DistanceType metric,
                                rmm::cuda_stream_view stream,
                                rmm::mr::device_memory_resource* mr)
 {
-  rmm::device_uvector<float> distances(n_rows * n_clusters, stream, mr);
-
-  float alpha;
-  float beta;
   switch (metric) {
-    case raft::distance::DistanceType::InnerProduct: {
-      alpha = -1.0;
-      beta  = 0.0;
-    } break;
     case raft::distance::DistanceType::L2Expanded:
-    case raft::distance::DistanceType::L2Unexpanded: {
-      rmm::device_uvector<float> sqsum_centers(n_clusters, stream, mr);
-      rmm::device_uvector<float> sqsum_data(n_rows, stream, mr);
-      utils::dots_along_rows(n_clusters, dim, centers, sqsum_centers.data(), stream);
-      utils::dots_along_rows(n_rows, dim, dataset, sqsum_data.data(), stream);
-      utils::outer_add(
-        sqsum_data.data(), n_rows, sqsum_centers.data(), n_clusters, distances.data(), stream);
-      alpha = -2.0;
-      beta  = 1.0;
-    } break;
-    // NB: update the description of `knn::ivf_flat::build` when adding here a new metric.
-    default: RAFT_FAIL("The chosen distance metric is not supported (%d)", int(metric));
+    case raft::distance::DistanceType::L2SqrtExpanded: {
+      // todo(lsugy): pass
+      auto workspace = raft::make_device_vector<char, IdxT>(handle, (sizeof(int)) * n_rows);
```

Comment:
We could make use of the memory_resoure argument (alternatively, one could expose an mr arg for make_device_vector).
```suggestion
      auto workspace = raft::make_device_mdarray<char, IdxT>(handle, mr, make_extents<IdxT>((sizeof(int)) * n_rows));
```